backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/movement/MoveHUCommand.java

```diff
@@ -0,0 +1,215 @@
+	private final IHUMovementBL huMovementBL = Services.get(IHUMovementBL.class);
+	private final IHandlingUnitsBL handlingUnitsBL = Services.get(IHandlingUnitsBL.class);
+	private final IWarehouseDAO warehouseDAO = Services.get(IWarehouseDAO.class);
+	private final HUTransformService huTransformService;
+	private final HUQRCodesService huQRCodesService;
+
+	private final List<HUIdAndQRCode> huIdAndQRCodes;
+	private final GlobalQRCode targetQRCode;
+
+	@Builder
+	private MoveHUCommand(
+			@NonNull final HUQRCodesService huQRCodesService,
+			@NonNull final List<HUIdAndQRCode> husToMove,
+			@NonNull final GlobalQRCode targetQRCode)
+	{
+		this.huQRCodesService = huQRCodesService;
+		this.huTransformService = HUTransformService.builder()
+				.huQRCodesService(huQRCodesService)
+				.build();
+
+		this.huIdAndQRCodes = husToMove;
+		this.targetQRCode = targetQRCode;
+
+	}
+
+	public void execute()
+	{
+		trxManager.runInThreadInheritedTrx(this::executeInTrx);
+	}
+
+	private void executeInTrx()
+	{
+		if (LocatorQRCode.isTypeMatching(targetQRCode))
+		{
+			final LocatorQRCode locatorQRCode = LocatorQRCode.ofGlobalQRCode(targetQRCode);
+
+			moveHUIdsToLocator(extractHUIdsToMove(), locatorQRCode.getLocatorId());
+		}
+		else if (HUQRCode.isTypeMatching(targetQRCode))
+		{
+			getMoveToTargetHUConsumer(HUQRCode.fromGlobalQRCode(targetQRCode))
+					.accept(extractHUIdsToMove());
+		}
+		else
+		{
+			throw new AdempiereException("Move target not handled: " + targetQRCode);
+		}
+	}
+
+	@NonNull
+	private Consumer<List<HuId>> getMoveToTargetHUConsumer(@NonNull final HUQRCode huqrCode)
+	{
+		final I_M_HU targetHU = handlingUnitsBL.getById(huQRCodesService.getHuIdByQRCode(huqrCode));
+		if (handlingUnitsBL.isDestroyed(targetHU))
+		{
+			handlingUnitsBL.setHUStatus(targetHU, PlainContextAware.newWithThreadInheritedTrx(), X_M_HU.HUSTATUS_Active);
+		}
+
+		if (handlingUnitsBL.isLoadingUnit(targetHU))
+		{
+			return getMoveToLUConsumer(targetHU);
+		}
+		else if (handlingUnitsBL.isTransportUnit(targetHU))
+		{
+			return getMoveToTUConsumer(targetHU);
+		}
+		else if (handlingUnitsBL.isVirtual(targetHU))
+		{
+			return getMoveToCUConsumer(targetHU);
+		}
+		else
+		{
+			throw new AdempiereException("Unsupported target!");
+		}
+	}
+
+	@NonNull
+	private Consumer<List<HuId>> getMoveToCUConsumer(final I_M_HU targetHU)
+	{
+		if (!handlingUnitsBL.isVirtual(targetHU))
+		{
+			throw new AdempiereException("Invalid target HU! Expecting CU type, but got=" + handlingUnitsBL.getHU_UnitType(targetHU))
+					.markAsUserValidationError();
+		}
+
+		final LocatorId locatorIdOfTargetHU = warehouseDAO.getLocatorIdByRepoId(targetHU.getM_Locator_ID());
+
+		return (huIdsToMove) -> {
+			final List<I_M_HU> husToMove = handlingUnitsBL.getByIds(huIdsToMove);
+
+			if (husToMove.stream().anyMatch(tu -> !handlingUnitsBL.isVirtual(tu)))
+			{
+				throw new AdempiereException("Expecting only CUs to be moved");
+			}
+
+			moveHUsToLocator(husToMove, locatorIdOfTargetHU);
```

Comment:
why do we have to move it to target locator before doing cusToExistingCU ?

Reply:
I've added it for consistency reasons, we do it for all the other movements, behind the scenes there is an M_Movement created, I imagine that has its purpose, not sure what's that exactly, but I've played it safe and did the same thing here.. do you see any reason why we shouldn't do it?

Reply:
agree